File: JAVA/Karatsuba.java
package karatsubaalgorithm;
import java.math.*;
import java.util.Scanner;

public class Karatsuba {
	public static BigInteger binaryToDecimal(String binary) {
        return new BigInteger(binary, 2);
    }
    // Convert decimal BigInteger to binary string
    public static String decimalToBinary(BigInteger decimal) {
        return decimal.toString(2);
    }
    // Karatsuba multiplication algorithm
    public static BigInteger karatsubaMultiply(BigInteger x, BigInteger y) {
        int n = Math.max(x.bitLength(), y.bitLength());
        if (n <= 3) {
            return x.multiply(y);
        }
        int m = (n + 1) / 2;
        // Split x and y into high and low parts
        BigInteger xHigh = x.shiftRight(m);
        BigInteger xLow = x.subtract(xHigh.shiftLeft(m));
        BigInteger yHigh = y.shiftRight(m);
        BigInteger yLow = y.subtract(yHigh.shiftLeft(m));
        // Recursively compute sub-products
        BigInteger a = karatsubaMultiply(xHigh, yHigh);
        BigInteger b = karatsubaMultiply(xLow, yLow);
        BigInteger c = karatsubaMultiply(xHigh.add(xLow), yHigh.add(yLow)).subtract(a).subtract(b);
        // Combine sub-products to get the final result
        return a.shiftLeft(2 * m).add(c.shiftLeft(m)).add(b);
    }
	public static void main(String[] args) {
		// TODO Auto-generated method stub
			Scanner sc = new Scanner(System.in);
			System.out.println("Enter first string: ");
		 	String A = sc.nextLine();
		 	System.out.println("Enter second string: ");
	      	String B = sc.nextLine();
	      	
	        // Convert binary strings to decimal integers
	        BigInteger decimalA = binaryToDecimal(A);
	        BigInteger decimalB = binaryToDecimal(B);
	        
	        // Multiply using Karatsuba
	        BigInteger product = karatsubaMultiply(decimalA, decimalB);
	        // Convert the result back to a binary string
	        String binaryResult = decimalToBinary(product);
	        System.out.println("Product in Decimal Value: " + product);
	        System.out.println("Product in Binary: " + binaryResult);
	}
}

File: CPP/karatsuba.cpp
#include <iostream>
#include <string>
#include <cmath>

using namespace std;

int binaryToDecimal(string binary) {
    int decimal = 0;
    int n = binary.length();
  
    for(int i = 0; i < n; i++) {
        if(binary[i] == '1') {
            decimal += pow(2, n - 1 - i);
        }
    }
  
    return decimal;
}

string decimalToBinary(int decimal) {
    string binary = "";
  
    while(decimal > 0) {
        binary = to_string(decimal % 2) + binary;
        decimal /= 2;
    }
  
    return binary;
}

int karatsubaAlgo(string A, string B) {
    int num1 = binaryToDecimal(A);
    int num2 = binaryToDecimal(B);
  
    int product = num1 * num2;
  
    return product;
}